sdk/python/feast/job_service.py

```diff
@@ -145,6 +167,54 @@ def GetJob(self, request, context):
         return GetJobResponse(job=self._job_to_proto(job))
 
 
+def run_control_loop():
+    """Runs control loop that continuously ensures that correct jobs are being run."""
+    logging.info(
+        "Feast Job Service is starting a control loop in a background thread..."
+    )
+    client = feast.Client()
+    while True:
+        # Map of job hash -> (project, table_name)
+        table_metadata: Dict[str, Tuple[str, str]] = dict()
+        # Job hashes that should exist
+        final_job_hashes: Set[str] = set()
+
+        for project in client.list_projects():
+            feature_tables = client.list_feature_tables(project)
+            for feature_table in feature_tables:
+                if feature_table.stream_source is not None:
+                    params = get_stream_to_online_ingestion_params(
+                        client, project, feature_table, []
+                    )
+                    job_hash = params.get_job_hash()
+                    final_job_hashes.add(job_hash)
+                    table_metadata[job_hash] = (project, feature_table.name)
+
+        jobs_by_hash = list_jobs_by_hash(client, include_terminated=False)
+        # Job hashes that currently exist
+        existing_job_hashes = set(jobs_by_hash.keys())
+
+        job_hashes_to_cancel = existing_job_hashes - final_job_hashes
+        job_hashes_to_start = final_job_hashes - existing_job_hashes
+
+        for job_hash in job_hashes_to_cancel:
+            job = jobs_by_hash[job_hash]
+            logging.info(
+                f"Cancelling a stream ingestion job with job_hash={job_hash} job_id={job.get_id()} status={job.get_status()}"
+            )
+            job.cancel()
+
+        for job_hash in job_hashes_to_start:
+            project, table_name = table_metadata[job_hash]
+            logging.info(
+                f"Starting a stream ingestion job for project={project}, table_name={table_name} with job_hash={job_hash}"
+            )
+            feature_table = client.get_feature_table(name=table_name, project=project)
+            start_stream_to_online_ingestion(client, project, feature_table, [])
+
+        time.sleep(1)
```

Comment:
One recurring issue I've seen with daemon threads is obscure bugs when the thread dies for whatever reason (unexpected exceptions an whatnot). The main process will happily chug along, and you may only notice days later that your control loop isn't running anymore, or not notice this at all.

Is there any robust way to avoid that here? Maybe something like `try ... finally: sys.exit(1)` ?

Reply:
yeah we need to keep it alive, good catch...